### DISCUSSION TITLE
Promise.all or axios.all - how to mock?

### DISCUSSION THREAD

**Discussion by jordanss1** [mswjs/msw#1500] [Q&A]
I saw the other discussion from November based around this but I don't think it answers my question.

I want to mock this:

     const artistAndAlbum = [
        `https://api.spotify.com/v1/artists/${id}`,
        `https://api.spotify.com/v1/artists/${id}/albums?include_groups=album&limit=50`,
        `https://api.spotify.com/v1/artists/${id}/top-tracks?market=US`,
      ];

       axios
           .all(
                artistAndAlbum.map((endpoint) =>
                  axios.get(endpoint, {
                    headers: {
                      Authorization: accessToken,
                    },
                  })
                )
              )
              .then((responses) => {
                  //use responses

Currently I only have this as the handler:

    export const artistAndAlbumHandler = (data) => {
      const artistAndAlbum = [
        `https://api.spotify.com/v1/artists/:id`,
        `https://api.spotify.com/v1/artists/:id/albums`,
        `https://api.spotify.com/v1/artists/:id/top-tracks`,
      ];
  
      rest.get(artistAndAlbum[0], (req, res, ctx) => {
        return res(ctx.status(200), ctx.body(data[0]));
      }),
      rest.get(artistAndAlbum[1], (req, res, ctx) => {
        return res(ctx.status(200), ctx.body(data[1]));
      }),
      rest.get(artistAndAlbum[2], (req, res, ctx) => {
        return res(ctx.status(200), ctx.body(data[2]));
      }),
    ];
  
The "data" parameter is an array that I created, which matches the response from the actual network requests.

When I implement it and I console log the response it's just a big jumble of a response body, which doesn't match created array. This means the response is going through because there is no network error but the handler is not configured properly.

What is a simple way to configure this to mock Promise.all?

Thank you

**Comment by mattcosta7** [comment]
@jordanss1 do you mean to use [ctx.text()](https://mswjs.io/docs/api/context/text) instead of [ctx.body](https://mswjs.io/docs/api/context/body).

The `promise.all` or `axios.all` is unrelated to msw. Each request is separately handled by msw.

ctx.body is a raw body, without content-type or anything set, so you'll have a hard time parsing it properly. ctx.text will set a content type text/plain, so you can properly parse it as text later.   If you intentd to send json instead, you'd want to call [ctx.json](https://mswjs.io/docs/api/context/json)

**Reply by kettanaito** [reply]
Not to mention that what they probably need is `ctx.json()` since 99% of the time we handle JSON data on the web realm. 

I agree with this: no matter how your calls are orchestrated in your app, approach them on a per-endpoint basis, which means declaring a request handler for each call in MSW. Really, just imagine your mocks from the server's perspective where it's not concerned about how you structure your calls.

**Reply by jordanss1** [reply]
Hey guys @kettanaito @mattcosta7 , that's weird because I have had 23 tests that pass from my previous handlers which use ctx.body() and simple arrays with no JSON.

    export const artistAndTrackHandlers = (data) => {
      return [
        rest.post("https://accounts.spotify.com/api/token", (req, res, ctx) => {
          return res(ctx.status(200, "success"));
        }),
    
        rest.get("https://api.spotify.com/v1/search", (req, res, ctx) => {
          return res(ctx.status(200), ctx.body(data));
        }),
      ];
    };

Here's one of the arrays which is the data argument

    export const songResults = {
      tracks: {
        items: [
          {
            name: "Test 1",
            external_urls: { spotify: "www.spotify.com" },
            album: {
              images: [0, 0, 0],
              album_type: "single",
              name: "Test 1",
              total_tracks: 0,
            },
            artists: [{ name: "Test 1" }],
            track_number: 1,
            duration_ms: 10000,
          },
          {
            name: "Test 2",
            external_urls: { spotify: "www.spotify.com" },
            album: {
              images: [0, 0, 0],
              album_type: "single",
              name: "Test 2",
              total_tracks: 0,
            },
            artists: [{ name: "Test 2" }],
            track_number: 1,
            duration_ms: 10000,
          },
          {
            name: "Test 3",
            external_urls: { spotify: "www.spotify.com" },
            album: {
              images: [0, 0, 0],
              album_type: "single",
              name: "Test 3",
              total_tracks: 0,
            },
            artists: [{ name: "Test 3" }],
            track_number: 1,
            duration_ms: 10000,
          },
        ],
      },
    };

When console logged, it seems like the handler is returning an array with the items I need but they are adding extra objects regarding "request: XMLHttpRequestOveride" which is not what I added in the ctx.body(). Every request property is not what I need, the other objects are what I need:

FAIL src/components/__tests__/SearchList.test.js
  ● Console

    console.log
      [
        {
          data: { data: [Object] },
          status: 200,
          statusText: 'OK',
          headers: { 'x-powered-by': 'msw' },
          config: {
            transitional: [Object],
            adapter: [Function: xhrAdapter],
            transformRequest: [Array],
            transformResponse: [Array],
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            maxBodyLength: -1,
            env: [Object],
            validateStatus: [Function: validateStatus],
            headers: [Object],
            method: 'get',
            url: 'https://api.spotify.com/v1/artists/1',
            data: undefined
          },
          request: XMLHttpRequestOverride {
            _events: [],
            log: [Function],
            UNSENT: 0,
            OPENED: 1,
            HEADERS_RECEIVED: 2,
            LOADING: 3,
            DONE: 4,
            onreadystatechange: null,
            onabort: [Function: handleAbort],
            onerror: [Function: handleError],
            onload: null,
            onloadend: [Function: onloadend],
            onloadstart: null,
            onprogress: null,
            ontimeout: [Function: handleTimeout],
            url: 'https://api.spotify.com/v1/artists/1',
            method: 'GET',
            readyState: 4,
            withCredentials: false,
            status: 200,
            statusText: 'OK',
            response: [Object],
            responseType: 'text',
            responseText: [Object],
            responseXML: null,
            responseURL: 'https://api.spotify.com/v1/artists/1',
            upload: {},
            timeout: 0,
            _requestHeaders: [HeadersPolyfill],
            _responseHeaders: [HeadersPolyfill],
            async: true,
            user: undefined,
            password: undefined
          }
        },
        {
          data: { data: [Object] },
          status: 200,
          statusText: 'OK',
          headers: { 'x-powered-by': 'msw' },
          config: {
            transitional: [Object],
            adapter: [Function: xhrAdapter],
            transformRequest: [Array],
            transformResponse: [Array],
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            maxBodyLength: -1,
            env: [Object],
            validateStatus: [Function: validateStatus],
            headers: [Object],
            method: 'get',
            url: 'https://api.spotify.com/v1/artists/1/albums?include_groups=album&limit=50',
            data: undefined
          },
          request: XMLHttpRequestOverride {
            _events: [],
            log: [Function],
            UNSENT: 0,
            OPENED: 1,
            HEADERS_RECEIVED: 2,
            LOADING: 3,
            DONE: 4,
            onreadystatechange: null,
            onabort: [Function: handleAbort],
            onerror: [Function: handleError],
            onload: null,
            onloadend: [Function: onloadend],
            onloadstart: null,
            onprogress: null,
            ontimeout: [Function: handleTimeout],
            url: 'https://api.spotify.com/v1/artists/1/albums?include_groups=album&limit=50',
            method: 'GET',
            readyState: 4,
            withCredentials: false,
            status: 200,
            statusText: 'OK',
            response: [Object],
            responseType: 'text',
            responseText: [Object],
            responseXML: null,
            responseURL: 'https://api.spotify.com/v1/artists/1/albums?include_groups=album&limit=50',
            upload: {},
            timeout: 0,
            _requestHeaders: [HeadersPolyfill],
            _responseHeaders: [HeadersPolyfill],
            async: true,
            user: undefined,
            password: undefined
          }
        },
        {
          data: { data: [Object] },
          status: 200,
          statusText: 'OK',
          headers: { 'x-powered-by': 'msw' },
          config: {
            transitional: [Object],
            adapter: [Function: xhrAdapter],
            transformRequest: [Array],
            transformResponse: [Array],
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            maxBodyLength: -1,
            env: [Object],
            validateStatus: [Function: validateStatus],
            headers: [Object],
            method: 'get',
            url: 'https://api.spotify.com/v1/artists/1/top-tracks?market=US',
            data: undefined
          },
          request: XMLHttpRequestOverride {
            _events: [],
            log: [Function],
            UNSENT: 0,
            OPENED: 1,
            HEADERS_RECEIVED: 2,
            LOADING: 3,
            DONE: 4,
            onreadystatechange: null,
            onabort: [Function: handleAbort],
            onerror: [Function: handleError],
            onload: null,
            onloadend: [Function: onloadend],
            onloadstart: null,
            onprogress: null,
            ontimeout: [Function: handleTimeout],
            url: 'https://api.spotify.com/v1/artists/1/top-tracks?market=US',
            method: 'GET',
            readyState: 4,
            withCredentials: false,
            status: 200,
            statusText: 'OK',
            response: [Object],
            responseType: 'text',
            responseText: [Object],
            responseXML: null,
            responseURL: 'https://api.spotify.com/v1/artists/1/top-tracks?market=US',
            upload: {},
            timeout: 0,
            _requestHeaders: [HeadersPolyfill],
            _responseHeaders: [HeadersPolyfill],
            async: true,
            user: undefined,
            password: undefined
          }
        }
      ]

Do you have any idea why those extra request properties have been added?

Thanks

**Reply by kettanaito** [reply]
> I have had 23 tests that pass from my previous handlers which use ctx.body() and simple arrays with no JSON.

This is very likely because your request client infers the response body content type. Don't rely on that, respond with valid HTTP responses all the time. 

> We expose `ctx.body()` so that you get more flexibility in constructing custom responses (e.g. the ones with vendor-specific headers that `ctx.json()` or other utilities cannot represent). You never really want to use `ctx.body()` without `ctx.set()`. 

You seem to be logging out the entire result of your request client, which may include references to the `request` performed. There's nothing wrong with that, it's an implementation detail you can ignore. Interceptors do not interfere with your response payload in any shape or form. What you're observing is not wrong, you're just printing a much higher-context info about the response. What you're interested in is the `data` value.

**Reply by jordanss1** [reply]
@kettanaito sorry guys, once again it was my silly mistake and my own implementation issue. I sorted it now. 

So would the proper way to do things be to create a normal array or object but when it's passed to the handler use ctx.json() on it?

Thanks

**Reply by kettanaito** [reply]
No worries!

You create handlers for the endpoints you need. This has little to do with how you make the requests (apart from the request URL). You may be fetching different requests in sequence, in parallel, or even requesting the same endpoint multiple times (let's say it returns different results). That remains up to you and your implementation. Just approach mocks with MSW from the server's perspective and you will be set to go.
